src/System.Net.Http.WinHttpHandler/src/System/Net/Http/WinHttpHandler.cs

```diff
@@ -0,0 +1,2164 @@
+                        // settings ourself using our WinInetProxyHelper object.
+                        _proxyHelper = new WinInetProxyHelper();
+                        _sessionHandle = Interop.WinHttp.WinHttpOpen(
+                            IntPtr.Zero,
+                            _proxyHelper.ManualSettingsOnly ? Interop.WinHttp.WINHTTP_ACCESS_TYPE_NAMED_PROXY : Interop.WinHttp.WINHTTP_ACCESS_TYPE_NO_PROXY,
+                            _proxyHelper.ManualSettingsOnly ? _proxyHelper.Proxy : Interop.WinHttp.WINHTTP_NO_PROXY_NAME,
+                            _proxyHelper.ManualSettingsOnly ? _proxyHelper.ProxyBypass : Interop.WinHttp.WINHTTP_NO_PROXY_BYPASS,
+                            0);
+                        if (_sessionHandle.IsInvalid)
+                        {
+                            throw new HttpRequestException(
+                                SR.net_http_client_execution_error,
+                                WinHttpException.CreateExceptionUsingLastError());
+                        }
+
+                        _sessionHandle.DangerousAddRef(ref ignore);
+                    }
+                }
+            }
+        }
+
+        private bool GetServerCredentialsFromCache(
+            Uri uri,
+            out uint serverAuthScheme,
+            out NetworkCredential serverCredentials)
+        {
+            serverAuthScheme = 0;
+            serverCredentials = null;
+
+            NetworkCredential cred = null;
+
+            lock (_credentialCacheLock)
+            {
+                foreach (uint authScheme in s_authSchemePriorityOrder)
+                {
+                    cred = _credentialCache.GetCredential(uri, s_authSchemeStringMapping[authScheme]);
+                    if (cred != null)
+                    {
+                        serverAuthScheme = authScheme;
+                        serverCredentials = cred;
+
+                        return true;
+                    }
+                }
+            }
+
+            return false;
+        }
+
+        private void SaveServerCredentialsToCache(Uri uri, uint authScheme, ICredentials serverCredentials)
+        {
+            string authType = s_authSchemeStringMapping[authScheme];
+            Debug.Assert(!string.IsNullOrEmpty(authType));
+
+            NetworkCredential cred = serverCredentials.GetCredential(uri, authType);
+            if (cred != null)
+            {
+                lock (_credentialCacheLock)
+                {
+                    try
+                    {
+                        _credentialCache.Add(uri, authType, cred);
+                    }
+                    catch (ArgumentException)
+                    {
+                        // The credential was already added.
+                    }
+                }
+            }
+        }
+
+        private async void StartRequest(object obj)
+        {
+            RequestState state = obj as RequestState;
```

Comment:
Could be ```(RequestState)obj``` instead of ```obj as RequestState```.  Since we require obj to be a RequestState at this point, that would help with diagnostics if this ever failed, as it'd resultin an InvalidCastException from here rather than a NullReferenceException from later in the method if obj was some other non-null type.